src/com/ms/silverking/cloud/topology/TopologyParser.java

```diff
@@ -11,157 +11,158 @@
+                  default:
+                    throw new RuntimeException("panic");
+                  }
+                  break;
+                case ENTRY:
+                  parentEntries.get(parentEntries.size() - 1).addChild(new TopologyEntry(token.getEntry()));
+                  break;
+                default:
+                  throw new RuntimeException("panic");
                 }
-            } while (line != null);
-            if (debug) {
-                System.out.println(root);
-            }
-            rootNode = TopologyEntry.entriesToNodes(root);
-            if (debug) {
-                System.out.println(rootNode.toStructuredString());
-            }
-            if (version == noVersion) {
-                return Topology.fromRoot(null, rootNode);
-            } else {
-                return Topology.fromRoot(null, rootNode, version);
+                prevToken = token;
+              }
             }
-        } catch (TopologyParseException tpe) {
-            throw new TopologyParseException(tpe.getMessage() +" lineNumber: "+ lineNumber);
-        } catch (IOException ioe) {
-            throw new IOException(ioe.getMessage() +" lineNumber: "+ lineNumber);
+          }
         }
+      } while (line != null);
+      if (debug) {
+        System.out.println(root);
+      }
+      rootNode = TopologyEntry.entriesToNodes(root);
+      if (debug) {
+        System.out.println(rootNode.toStructuredString());
+      }
+      if (version == noVersion) {
+        return Topology.fromRoot(null, rootNode);
+      } else {
+        return Topology.fromRoot(null, rootNode, version);
+      }
+    } catch (TopologyParseException tpe) {
+      throw new TopologyParseException(tpe.getMessage() + " lineNumber: " + lineNumber);
+    } catch (IOException ioe) {
+      throw new IOException(ioe.getMessage() + " lineNumber: " + lineNumber);
     }
-    
-    public static Topology parse(InputStream in) throws IOException, TopologyParseException {
-        return _parse(in, noVersion);
-    }
-    
-    public static Topology parseVersioned(File topoFile, long version) 
-                                            throws IOException {
-        return _parse(new FileInputStream(topoFile), version);
-    }
-    
-            
-    public static Topology parseVersioned(String def, long version) 
-                                            throws IOException {
-        return _parse(new ByteArrayInputStream(def.getBytes()), version);
-    }
+  }
 
-    private static void sanityCheckLine(String line) throws IOException {
-        if (line.indexOf(' ') >= 0) {
-            throw new IOException("Spaces disallowed: "+ line);
-        }
+  public static Topology parse(InputStream in) throws IOException, TopologyParseException {
+    return _parse(in, noVersion);
+  }
+
+  public static Topology parseVersioned(File topoFile, long version) throws IOException {
+    return _parse(new FileInputStream(topoFile), version);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the resource created by the following constructor call: *FileInputStream*. Currently, there are execution paths that do not contain closure statements. Either a) close the object returned by *FileInputStream()* in a try-finally block or b) close the resource by declaring the object returned by *FileInputStream()* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).